integration-tests/src/tests/client/features/chunk_validation.rs

```diff
@@ -137,47 +164,95 @@ fn test_chunk_validation_basic() {
             "Producing block at height {} by {}", tip.height + 1, block_producer
         );
         let block = env.client(&block_producer).produce_block(tip.height + 1).unwrap().unwrap();
-        if round > 1 {
-            for i in 0..num_shards {
-                let chunks = block.chunks();
-                let chunk = chunks.get(i).unwrap();
-                assert!(chunk.is_new_chunk(block.header().height()));
-            }
-        }
 
         // Apply the block.
         for i in 0..env.clients.len() {
+            let validator_id = env.get_client_id(i);
             tracing::debug!(
                 target: "chunk_validation",
-                "Applying block at height {} at {}", block.header().height(), env.get_client_id(i)
+                "Applying block at height {} at {}", block.header().height(), validator_id
             );
-            let blocks_processed =
-                env.clients[i].process_block_test(block.clone().into(), Provenance::NONE).unwrap();
+            let blocks_processed = if rng.gen_bool(prob_missing_chunk) {
+                env.clients[i]
+                    .process_block_test_no_produce_chunk(block.clone().into(), Provenance::NONE)
+                    .unwrap()
+            } else {
+                env.clients[i].process_block_test(block.clone().into(), Provenance::NONE).unwrap()
+            };
             assert_eq!(blocks_processed, vec![*block.hash()]);
         }
 
         env.process_partial_encoded_chunks();
         for j in 0..env.clients.len() {
             env.process_shards_manager_responses_and_finish_processing_blocks(j);
         }
-        env.propagate_chunk_state_witnesses();
+        let result = env.propagate_chunk_state_witnesses();
+        found_different_post_state_roots |= result.found_different_post_state_roots;
     }
 
+    // Check that at least one tx was fully executed, ensuring that executing
+    // state witness against non-trivial recorded storage is checked.
+    let mut has_executed_txs = false;
     for tx_hash in tx_hashes {
-        let outcome = env.clients[0].chain.get_final_transaction_result(&tx_hash).unwrap();
-        assert_matches!(outcome.status, FinalExecutionStatus::SuccessValue(_));
+        let outcome = env.clients[0].chain.get_final_transaction_result(&tx_hash);
+        if let Ok(outcome) = outcome {
+            if let FinalExecutionStatus::SuccessValue(_) = outcome.status {
+                has_executed_txs = true;
+            }
+        }
+    }
+    assert!(has_executed_txs);
+
+    // If probability of missing chunk is high, claim that it is enough for
+    // some chunk at epoch boundary to miss, likely causing two different post
+    // state roots in some state witness.
+    if prob_missing_chunk > 0.7 {
+        assert!(found_different_post_state_roots);
```

Comment:
How confident are you that this won't flake?

Reply:
We are fixing the seeds, should be fine

Reply:
I could have been more accurate with this...
Well, the actual probability passed here is 0.8. We have 4 epoch boundaries on each of 4 shards. So the probability of flake is `pow(0.2, 16)` which is around `10**-12`. I will add this to comment